### ISSUE TITLE
Vortex & MO2 Not Handling NXM Links

### ISSUE THREAD

**Issue by SentaiBrad** [sonic2kk/steamtinkerlaunch#637]
## System Information
- SteamTinkerLaunch version: Latest GIT 
- Distribution: SteamOS 
- Installation Method: ProtonUp-Qt

## Issue Description

This is a combination of one issue, that the wiki says STL solves, but for me, it does not. But to be clear, the bug inherently lies in the fact that Vortex is running through a Wine layer.

I had to re-read the Wiki a few times to make sure I wasn't crazy, so this MIGHT be a bug with STL also.

> "create and update the nxm-protocol mime type for directly opening NXM with Vortex (renaming that from Mod-Organizer 2 when found)"

When Vortex launched for me the first time, there was an error and it closed. I opened it again, and Vortex asked me if I wanted to "repair" the installation, and I did. It asked me if I wanted to set Vortex to open NXM links, and I did. However, when clicking on "Download with Manager" on Nexus, Firefox prompted me to pick an application and nothing more. Nothing Vortex or STL related was selectable. 

I have been helping a friend, trying to get Vortex up and running so they can use the Collections feature, but even for them clicking the "Download with Manager" button did nothing. They are also using a Steam Deck with ProtonUp-Qt, and never reported errors for their install, but I will focus on mine for this report. I just wanted to make sure to note this, as this is why I started testing for it.

As for solving the issue: I presume that STL is running a script (like the other vortex and mo2 installers) to enable the NXM link manually. Assuming that the above bug report is valid, but the issue is not in relation to this script and the script works, it's more of a Feature Request to get this added to the GUI so that a user can have the script re-apply the NXM link. However, I am unaware if this is an issue with the script, or that it just needs to be run again, so I will Bug Report it just in case.

Also worth noting: In one MO2 testing instance for me (before STL), providing my API key manually worked flawlessly. For Vortex, we're not given an API key option as it's a tool by Nexus, for Nexus. They prompt users to log in to check things like Premium, endorse files, etc. Unsure if they store the API key locally, but they provide no way for the user to input this via the GUI. If they do store the API locally, it might be feasible to add an option to the STL GUI for a user to copy and paste their Nexus API key for it to be inserted manually by STL into Vortex. That is a passing thought though, I have not dug in far enough to see if they store it at all, or purely just use the login method dynamically. 

Side note: on triple checking my information and Launching Vortex again, it looks like it KEEPS asking me to repair the installation. Unsure if this is because of the Wine Layer and if I can Ignore it, or if I need to repair it. "Repairing" the installation might be why the NXM link no longer works.

![image](https://user-images.githubusercontent.com/26223818/197424849-e4338113-3ed8-450e-878a-950dc7afd43a.png)

## Logs
The only .log I have in the "dev/shm/steamtinkerlaunch/steamtinkerlaunch" path is winres.log. Says this twice:

> Sun Oct 23 05:00:50 PM PDT 2022 INFO - pollWinRes - Using GEOM '--geometry=1280x800+0+0' from '/home/deck/.config/steamtinkerlaunch/guicfgs/1280x800/22320/SteamTinkerLaunch-PrepareVortexGameStatus.conf'

**Comment by sonic2kk** [comment]
I'm not sure why Vortex isn't working but as far as I know it has been barely tested on Steam Deck. There is currently an open, pinned issue discussing Vortex-related problems. You should have a look through that issue and see if there's anything that answers the problems you are having.

I don't use Vortex personally, sorry, but the NXM link handling should be working, but it might not be working on Steam Deck. I'm not totally sure, you might want to ask in the dedicated Vortex issue to bring more attention to it.

**Comment by sonic2kk** [comment]
One thing I will mention is, NXM link handling might not be working if you're launching Vortex from the Main Menu extra button, or from the command line with `steamtinkerlaunch vortex start`.

If Vortex has a "GUI" wait requester mode like MO2 does, you might have to use that to get link handling to work.

But again, reading through and asking on the Vortex issue might give you some more hints. The Vortex side of things is one of the areas of STL I've touched the least.

**Comment by sonic2kk** [comment]
From skimming the more recent comments in the Vortex issue it seems like there might be some improvements needed.

Not sure when I'll get time or motivation to look into this as I am very inexperienced in this area, but contributions would be welcome to improve Vortex support

**Comment by SentaiBrad** [comment]
While doing some testing for issues #638 #635 #639 the "Download with Manager" option on Nexus, in Firefox, does nothing for either manager.

**Comment by sonic2kk** [comment]
For MO2, are you running in MO2 Game Mode (pressing the STL "Play" button) or Standalone Mode (Main Menu button)? If you're running MO2 in Standalone Mode, this is known behaviour and is fixed I believe in the MO2 rewrite branch which was created by Frostworx. I haven't looked much into that branch, but I know it has some improvements for MO2.

However if MO2 isn't working for you in Game Mode, there might be a problem here...

**Comment by SentaiBrad** [comment]
> For MO2, are you running in MO2 Game Mode (pressing the STL "Play" button) or Standalone Mode (Main Menu button)? If you're running MO2 in Standalone Mode, this is known behaviour and is fixed I believe in the MO2 rewrite branch which was created by Frostworx. I haven't looked much into that branch, but I know it has some improvements for MO2.
> 
> However if MO2 isn't working for you in Game Mode, there might be a problem here...

MO2 and Vortex overall work except for loading mods from Nexus. MO2 *WAS* working when I originally made this thread and was testing Vortex. I put my API in manually as it wouldn't launch to log me in. Sometime later, I{ was downloading a patch and it didn't work anymore. Launching Nexus to log me in work, but downloads still didn't. Cleared this, and re-entered API no go.

I just wrote all that to also again realize, this is still most likely nothing STL can solve lmao. I only lean that way cause it *did* at one point. I saw that re-write mentioned. I am excited.

**Comment by sonic2kk** [comment]
How are you launching MO2 though? Are you doing it through MO2 Game Mode or Standalone Mode? If you aren't sure what this means, refer to the [MO2 wiki page](https://github.com/sonic2kk/steamtinkerlaunch/wiki/Mod-Organizer-2) for more detail.

If you're launching from Standalone Mode, then it is expected that MO2 won't work when downloading from Nexus. For Vortex, I honestly don't know unfortunately.

**Comment by SentaiBrad** [comment]
GameMode. I click play on New Vegas, go through the first GUI, wait for the "Start MO2" or "No MO2", click Start MO2 and it launches from there.

**Comment by sonic2kk** [comment]
Hmm, I'm not sure why it has stopped working then to be honest. It's worked for me in the past and as mentioned it worked for you as well. Maybe restart your Steam Deck?

**Comment by SentaiBrad** [comment]
> Hmm, I'm not sure why it has stopped working then to be honest. It's worked for me in the past and as mentioned it worked for you as well. Maybe restart your Steam Deck?

I have. But, its not the end of the world for me. A very easy workaround is to just download the file manually and download it to (or copy it to) the MO2 downloads folder. MO2 see's it right away.

**Comment by sonic2kk** [comment]
> A very easy workaround is to just download the file manually

Yes, this is also my preferred method of installing mods, generally, especially larger modpacks.

But aside from that, the button should still be working and I am not totally sure why it would be broken. I'm not sure how to debug further if it just randomly stopped working.

You could check the steps outlined in [this section of the MO2 page](https://github.com/sonic2kk/steamtinkerlaunch/wiki/Mod-Organizer-2#mod-manager-button-on-nexus-mods-isnt-working), but I am not sure how much luck you'll have - I think in one reported case this worked, but I also equally think I remember a case where this did not resolve it (though I think for the latter case, the user wasn't using MO2 Game Mode).

**Comment by SentaiBrad** [comment]
> > A very easy workaround is to just download the file manually
> 
> Yes, this is also my preferred method of installing mods, generally, especially larger modpacks.
> 
> But aside from that, the button should still be working and I am not totally sure why it would be broken. I'm not sure how to debug further if it just randomly stopped working.
> 
> You could check the steps outlined in [this section of the MO2 page](https://github.com/sonic2kk/steamtinkerlaunch/wiki/Mod-Organizer-2#mod-manager-button-on-nexus-mods-isnt-working), but I am not sure how much luck you'll have - I think in one reported case this worked, but I also equally think I remember a case where this did not resolve it (though I think for the latter case, the user wasn't using MO2 Game Mode).

Frankly, if Linux Gaming is going to get a big push, we need native tools. But hey, I can dream :D 

Super excited for the re-write.

**Comment by sonic2kk** [comment]
> Frankly, if Linux Gaming is going to get a big push, we need native tools. But hey, I can dream :D

I'm right there with you, and Vortex Mod Manager may have a non-zero chance of getting a native port in the future (totally external to STL). But there are questions around how it will integrate with Wine/Proton and other related questions, *if* it happens at all.

But STL does its best to bridge the gap. The NXM link handler has been a bit of a pain-point recently with a couple of issues, but I believe they were resolved in the rewrite(?). There was some talk about a similar solution being implemented for Vortex.

The MO2 branch changes are pending further testing (I wouldn't recommend testing stuff like this on Deck, I would do it on a PC if possible). ProtonUp-Qt's STL-Git download lets you choose a branch (untested on Deck, recommended to do this on PC) so you could install and test the rewrite branch and give feedback at #598.

cc @frostworx since he is the one that made those changes, and iirc they are just pending further testing and possibly some conflict cleanup :-)

**Comment by SentaiBrad** [comment]
>so you could install and test the rewrite branch and give feedback at #598.
> 
> cc @frostworx since he is the one that made those changes, and iirc they are just pending further testing and possibly some conflict cleanup :-)

Did someone say testing... my ears were burning.

**Comment by SentaiBrad** [comment]
> The MO2 branch changes are pending further testing (I wouldn't recommend testing stuff like this on Deck, I would do it on a PC if possible). ProtonUp-Qt's STL-Git download lets you choose a branch (untested on Deck, recommended to do this on PC) so you could install and test the rewrite branch and give feedback at #598.

Excuse the noobiness, but I don't see any Git info for STL in ProtonUp. 

![image](https://user-images.githubusercontent.com/26223818/197613339-b446e7ba-77eb-421c-815f-3de1ea729be2.png)

![image](https://user-images.githubusercontent.com/26223818/197613121-861adf23-ef62-4887-9526-56ca09eb2b1c.png)

The bottom picture just scrolls to progressively older versions.

**Comment by sonic2kk** [comment]
> Excuse the noobiness, but I don't see any Git info for STL in ProtonUp.

Ah, you need to enable I think it's called "Advanced Mode" it in the "About" section - It enables Git versions of some tools including SteamTinkerLaunch

Also, if you do end up testing:
1. Unless you're very brave you might want to avoid doing it on Steam Deck
2. Backwards compatibility with existing mod installations is not guaranteed, *but* I think it should maybe work okay (it'll be merged soon-ish hopefully so we need to find out eventually if it does break existing installs)
3. Feedback should be given in #598 :-)

**Comment by SentaiBrad** [comment]
> > Excuse the noobiness, but I don't see any Git info for STL in ProtonUp.
> 
> Ah, you need to enable I think it's called "Advanced Mode" it in the "About" section - It enables Git versions of some tools including SteamTinkerLaunch
> 
> Also, if you do end up testing:
> 
>     1. Unless you're very brave you might want to avoid doing it on Steam Deck
> 
>     2. Backwards compatibility with existing mod installations is not guaranteed, _but_ I think it should maybe work okay (it'll be merged soon-ish hopefully so we need to find out eventually if it does break existing installs)
> 
>     3. Feedback should be given in [⚠ MO2 - Mod Organizer 2 overhaul discussion  #598](https://github.com/sonic2kk/steamtinkerlaunch/issues/598) :-)

I don't want to get the Laptop out right now. ProtonUp-Qt is still getting stuck at 99% though.

**Comment by sonic2kk** [comment]
I'm not sure why ProtonUp-Qt is hanging, but it'll be investigated as part of the work being done to improve the STL install experience with ProtonUp-Qt (and some improvements around how it handles dialogs). Might be fixed in the current git of ProtonUp-Qt and those changes just haven't made their way into an AppImage/Flatpak.

> I don't want to get the Laptop out right now.

Very fair, I do STL dev work on my Deck so I can hardly talk, but that is my "disclaimer" out of  the way :sweat_smile:

**Comment by SentaiBrad** [comment]
What....?

![image](https://user-images.githubusercontent.com/26223818/197621044-45834e2a-652c-4aeb-8dc5-abbd4c22c1fa.png)

Also, when I look at the script for installing STL, I clearly see:
PROGVERS="v11.12.20221024-5"

But once it's installed 
![image](https://user-images.githubusercontent.com/26223818/197621946-09cfd5ae-f959-47db-ae90-8d0d3c2db6de.png)

**Comment by sonic2kk** [comment]
I'm not sure what the problem is here.

The correct command to clone a new repository is to pass it the URL, e.g. `git clone https://github.com/sonic2kk/steamtinkerlaunch.git`

I'm not sure what's going on with the last command but that version is very old. I'm not sure what is going on here exactly.

**Comment by SentaiBrad** [comment]
> I'm not sure what the problem is here.
> 
> The correct command to clone a new repository is to pass it the URL, e.g. `git clone https://github.com/sonic2kk/steamtinkerlaunch.git`
> 
> I'm not sure what's going on with the last command but that version is very old. I'm not sure what is going on here exactly.

I was super out of it, getting no sleep over last night, so Ignore me there. I'll test when I am able here in a bit.

**Comment by sonic2kk** [comment]
Heh, I know the realm of sleep deprivation all too well so no worries :-) No rush!

**Comment by frostworx** [comment]
no need to cc me, as I'm still auto-subscribed to most issues.
I simply don't have much time to even read all the incoming mails :)
iirc the nxm handler rewrite worked fine for me both for vortex and mo2, and I was expecting some people to actually test it, because I do not use any of them.
Feedback was low, so I did all later mod unrelated changes in that branch as well to avoid having to waste even more time into that by merging everything later.
Maybe some people contribute something to the still open issues, else merging the corresponding changes from the branch (manually) and let new logs come in might be the easiest approach for the current situation.
I won't find any time to work on this _(one of the reasons I gave the project away)_ and to be honest, I would not even find the motivation anymore until I'm not interested in the feature personally _(another reason I gave the project away)_

**Comment by sonic2kk** [comment]
No problem, sorry for the ping. I just didn't want to potentially be talking nonsense about the mo2-overhaul branch :-)

> Maybe some people contribute something to the still open issues, else merging the corresponding changes from the branch (manually) and let new logs come in might be the easiest approach for the current situation.

I would agree. I'm hoping we can get more people to test it and I'm planning to test the branch with my existing mod installs soon-ish, just once I have a bit more free time from work to spend a long weekend testing and troubleshooting.

Just wondering, aside from potential conflicts and assuming it is mostly bug-free, is there anything outstanding you wanted to/want to add before a merge? It'll probably be a couple of weeks at least but I'm just trying to get an idea of how "merge ready" it is. So basically outside of testing, is there anything holding it back from being merged?

Thanks for responding either way, and like I saw in a few other issues you are always welcome to chip in whenever you have time :slightly_smiling_face:

**Comment by frostworx** [comment]
iirc there's nothing else which would require further tests.
I wish you all the best with attracting new testers :)

**Comment by SentaiBrad** [comment]
> no need to cc me, as I'm still auto-subscribed to most issues. I simply don't have much time to even read all the incoming mails :) iirc the nxm handler rewrite worked fine for me both for vortex and mo2, and I was expecting some people to actually test it, because I do not use any of them. Feedback was low, so I did all later mod unrelated changes in that branch as well to avoid having to waste even more time into that by merging everything later. Maybe some people contribute something to the still open issues, else merging the corresponding changes from the branch (manually) and let new logs come in might be the easiest approach for the current situation. I won't find any time to work on this _(one of the reasons I gave the project away)_ and to be honest, I would not even find the motivation anymore until I'm not interested in the feature personally _(another reason I gave the project away)_

Having personal reasons to step away from a project, even if you love it, is the more adult and mental health-conscious thing you could do. I appreciate the time and effort you put into this project, and then making sure someone was capable of taking over when you couldn't. <3

> No problem, sorry for the ping. I just didn't want to potentially be talking nonsense about the mo2-overhaul branch :-)
> 
> > Maybe some people contribute something to the still open issues, else merging the corresponding changes from the branch (manually) and let new logs come in might be the easiest approach for the current situation.
> 
> I would agree. I'm hoping we can get more people to test it and I'm planning to test the branch with my existing mod installs soon-ish, just once I have a bit more free time from work to spend a long weekend testing and troubleshooting.
> 
> Just wondering, aside from potential conflicts and assuming it is mostly bug-free, is there anything outstanding you wanted to/want to add before a merge? It'll probably be a couple of weeks at least but I'm just trying to get an idea of how "merge ready" it is. So basically outside of testing, is there anything holding it back from being merged?
> 
> Thanks for responding either way, and like I saw in a few other issues you are always welcome to chip in whenever you have time 🙂

If the rewrite fixes this for sure, then I'm not gonna test any further with the current builds. When it's rolled into the main branch, or if you still want me to grab that branch and test separately then I 100% will re-test.

**Comment by sonic2kk** [comment]
> if you still want me to grab that branch and test separately then I 100% will re-test.

There is no pressure but if you are able and willing, *all* feedback on the current state of the rewrite and things that don't work would be much appreciated! Any and all testing before I test and merge would be awesome :smile:

**Comment by sonic2kk** [comment]
I'm really curious as to why this doesn't work. If anyone wants to test the `mo2-overhaul` branch on their Deck to see if this fixes their issue with MO2, you can switch to using STL-git if you aren't already, and then run  `git checkout mo2-overhaul`, then re-install SteamTinkerLaunch by running `./steamtinkerlaunch`, that will install a version of STL with the MO2 version (the version number should end in `20221014`, then you'll know you're using the right version).

The way STL handles launching and manages MO2 is different in the rewrite branch, and there are some things in it that make me suspect it would fix the downloading issue. Firefox Flatpak on the Deck shouldn't be causing the problem as I tested on my laptop and it was fine. I suspect this _could_ be a Wine problem but I am not totally sure - Frostworx noted in the rewrite branch and associated MO2 overhaul testing issue that Wabbajack did not show a prompt to authenticate with Nexus.

I am not sure if much can be done from the STL side on this, but I will look into it and investigate on my PC when I am testing the rewrite branch.

**Comment by har-nick** [comment]
I'm sorry if this isn't the correct thread to be putting this, but I'm currently experiencing the same issue.

SteamTinkerLaunch is running as a Compatibility Tool through Steam.

<details>
<summary>Like so:</summary>

![image](https://user-images.githubusercontent.com/56137506/200989100-cdbee4a1-e2cf-46ff-8aa5-2bfc28811a20.png)
</details>

Mod Organizer 2 is enabled in GUI mode in Skyrim's Game Menu

<details>
<summary>Like so:</summary>

![image](https://user-images.githubusercontent.com/56137506/200991474-532b72c4-fb4a-4114-b273-3834c35525a6.png)
</details>

After launching Skyrim, waiting through the Game Menu prompt, and selecting MO2 GUI on the MO2 prompt, MO2 launches as normal.

<details>
<summary>A notification is sent through:</summary>

![image](https://user-images.githubusercontent.com/56137506/200992091-3b31c014-cf12-4bf6-9f94-0f006c8c01cf.png)

(Ignore the faulty updating notification. Fixed [here](https://github.com/sonic2kk/steamtinkerlaunch/issues/648#issuecomment-1309716794))
</details>

Clicking the Mod Manager Download button on the Nexus has no visual effect.

But like the other user posted, MO2 catches the request and downloads the mod as normal if I click the Mod Manager Download button without launching MO2 first.

Checked the link you sent regarding the Desktop file and mine is as shown in the example.

Would you mind letting me know what logs I should send over, if they would be of help?

**Comment by sonic2kk** [comment]
Hi there!

Please keep in mind that STL's MO2 link handling currently when MO2 is in Standalone Mode (from the Main Menu extra button). If you start in Game Mode (when you press STL's "Play" button), then link handling is known to not work. This may be fixed in the mo2-rewrite branch

If you are already doing that, there is no fix currently for the link handling. OP here reported that it suddenly stopped working for them with both Vortex and MO2 on Steam Deck. You'll just have to download your mods manually for now in that case. The MO2-rewrite branch may fix the issue but no guarantees.

If the MO2 rewrite branch doesn't fix this issue, then I'm afraid you're out of luck :frowning_face: 

P.S. glad you got the Proton issue fixed :smile:

**Comment by sonic2kk** [comment]
By the sounds of it, the MO2 portion of this issue may be solved in the mo2-overhaul branch. I wonder if whatever fixes are in place for MO2 could also be put in place for Vortex :thinking:

**Comment by sonic2kk** [comment]
As HedgeModManager is having the same issue on SteamOS, but `xdg-open` works (and I suspect it also works for NXM links too), I think this is a SteamOS bug. The `mo2-overhaul` branch also seems to fix MO2 links.

Vortex links seem to be hit and miss, but more testing once SteamOS 3.4 is out would be appreciated.

**Comment by har-nick** [comment]
The issues I had were while using Kwin. Maybe it's a Wayland thing?

**Comment by sonic2kk** [comment]
I'm on Plasma Wayland and MO2 nxm links work fine for me

**Comment by har-nick** [comment]
Interesting. Then I really couldn't say more, sorry.

**Comment by sonic2kk** [comment]
The MO2-overhaul branch fixed the issue for you though, correct? :slightly_smiling_face:

**Comment by har-nick** [comment]
That's right.

**Comment by sonic2kk** [comment]
I'll have to actually check what changed in that branch and see if any of it can be applied to Vortex, but likely any Vortex fixes won't be implemented before v12. The `mo2-overhaul` branch is a priority to merge before release though. Once I have some time I'll look into it again (been absolutely *wiped* after work this week)

**Comment by sonic2kk** [comment]
On Steam Deck, I have the same link handling problem with Hedge Mod Manager. It seems as though Firefox won't call the `.desktop` file for some reason. It never gets executed on Steam Deck and I am not sure why.

On my PC and laptop, with regular Firefox and with Firefox Flatpak, it works fine. I'll have to investigate into why it isn't working on Steam Deck but I suspect the issues on Steam Deck of link handling not working is related to this.

**Comment by jhrldev** [comment]
I'm having the same issue, tho, `mo2-overhaul` may have fixed it, I need to test.

In my case, it doesn't seem to be a Firefox issue at all, if I run the command directly on console (inside the right working directory), nothing happens at all. Commands get appended to `cmdline.txt` (idk the purpose of this file, but it's a signal STL is receiving the command correctly) but MO2 never starts downloading the mod.

Aside from that, I want to highlight some things.

> Frankly, if Linux Gaming is going to get a big push, we need native tools. But hey, I can dream :D

Yes, I always thought of doing that on my own, I just don't how long it would take and how the community would engage on this kind of project, because we need also to be aware that there are no standard way of installing mods.

The way we do for SK:AE (and Skyrim in general) for example, may be similar to how we do to Fallout 4 and probably differ a little for the future Starfield game, but certainly not the same as for TW3 and CP:77, MO2 supports 50+ games and Vortex way more, it would take some time to have all working. It obviously would be a thing built from the ground, which means, a lot of work to be done. Vortex is likely and Electron app, but still relies a lot on Windows APIs to be simple to port.

> [...] But there are questions around how it will integrate with Wine/Proton and other related questions, if it happens at all. [...]

AFAIK, from both my coding and modding experience (and Linux experience as well, it had been more than 10 years working with all of those things), this shouldn't matter that much, it's more a thing of copying things around and tweaking some files, so as long as we have the game path, this should be easy. If we ever encounter something like file patching, which sometimes have *Windows only tools*, we have both the choice of writing your own alternative, or just relying on the Windows version through Wine, this would be way less of a nightmare than trying to make the whole thing work from Wine/Proton (like MO2 and Vortex).

Some games may need additional dependencies (like .NET Framework), but most of the time those are already there, or we can just call `protontricks` or `winetricks` to deal with it, but I never faced this scenario on Windows either.

**Update:** It seems that `mo2-overhaul` fixed it, at least for Firefox, everything is working well there.

The only thing annoying me on the matter is that opening MO2 or Vortex completely resets the K&M driver and draw them unusable for 2-7 seconds (sometimes only the keyboard, others only the mouse, and after a couple of runs, both are resetted), probably an issue with *OpenRazer* module, reconnecting doesn't seem to make any difference, they still not responsive for some seconds. Peripherals that are not using *openrazer* module are not affected.

**Comment by sonic2kk** [comment]
Woohoo, more testing for the mo2-overhaul branch, thank you!

I agree with your comments on native tools and how they could be implemented. The only thing I would add is that newer versions of .NET (v6 and above I believe) have what I understand to be reasonably good native Linux support. Of course, a significant amount of tools (HMM, MO2) are still made with dotnet4 in mind, but my point here is that going forward, tools that rely on .NET should be a bit easier to port to Linux and/or should rely less on Winetricks. Vortex, for example, uses dotnet6, which STL downloads and runs the installer for manually as it works with newer Wine versions.

> The only thing annoying me on the matter is that opening MO2 or Vortex completely resets the K&M driver and draw them unusable for 2-7 seconds (sometimes only the keyboard, others only the mouse, and after a couple of runs, both are resetted), probably an issue with OpenRazer module, reconnecting doesn't seem to make any difference, they still not responsive for some seconds. Peripherals that are not using openrazer module are not affected.

Sorry you're running into this. I hate to give responses like this but afaik there isn't anything STL is going here to break this module. I'm not sure what might be causing this. SteamTinkerLaunch is quite a hefty Bash file but if you want to have a look at the `startMO2` code and `startVortex` code there shouldn't really be anything here that breaks it. The only *potential* thing I can see is with Vortex and its `LD_LIBRARY_PATH` and `LD_PRELOAD` variables being emptied before running. But I think it's slim that this would be causing issues, plus this issue affects Vortex and MO2 whereas this variable emptying only happens with Vortex.

Still, implementing some hardcoded workarounds specifically for this would be useful I think. I'm not sure what would be needed but if we can find out a fix (maybe setting a variable or something manually before startup?) I wouldn't mind implementing something to work around this issue :-)

<hr>

Thanks for the discussion here, it's much appreciated!

**Comment by sonic2kk** [comment]
Just had a look at the code. The function that manages MO2 links is the same one that manages Vortex links. So if the MO2 branch fixes link handling for MO2, it may also fix Vortex links.

**Comment by sonic2kk** [comment]
The MO2 overhaul branch has been merged. I have not tested Vortex since the merge but MO2 link handling worked in a brief test I did (though it always worked for me :sweat_smile:). I did not test on Steam Deck though.

Further testing here following this merge would be appreciated from actual Vortex users!

**Comment by Plamper** [comment]
Hi, the link handling seems to be still broken on SteamOS 3.4. I managed to make it work in a very cursed kind of way. Everytime I would click a nxm-link a kde application chooser would show up. There was no way to choose MO2 tho. Then I removed the `NoDisplay=true` line from the .desktop file in `~/.local/share/applications`. After that I was able to choose MO2 from the application chooser and all opened links after that would open in MO2.

**Comment by sonic2kk** [comment]
That is a very interesting hint, thank you for looking into that! I think experimentally then we could set `NoDisplay=true` when we create the .desktop file. I'll probably also set this for Hedge Mod Manager.

On Desktop, all this should really do is stop an icon from showing up beside the cursor. But if there's no icon set I just figured it made no difference.

I'll mess around with this and see if it breaks anything on desktop or has any undesired effects, but this could fix some of the frustrating behaviours especially on SteamOS where for some reason it forces you into using the annoying portal. On Desktop on my PC and laptop at least even in Firefox Flatpak it hasn't been an issue for me.

Also, if you're looking for a slightly less hacky workaround, pasting an NXM link (the one attached to the Mod Manager button, *not* the Nexusmods page URL) into the command `xdg-open` should also work to open NXM links.

<hr>

Thanks for finding this out, once I have some free time I'll see if this has any breaking / undesirable effects. The most undesirable effect would be if the link handlers started showing up in the application launcher.

**Comment by sonic2kk** [comment]
Just tested and this does have the annoying effect of making the link handler show up in the application list.

A different solution will be needed here then.

I wonder, what if you set the following in the `.desktop` file:

```
NoDisplay=false
Hidden=true
```

On my desktop with Plasma Wayland 5.26.4 and Arch Linux this at least keeps the desktop entry hidden but I'm not sure if it preserves the link handling yet. And I don't mod games on Steam Deck so I can't test SteamOS

**Comment by InsaneJ** [comment]
Not sure if this is related. I'm getting system tray popup message:
File 'X' already downloaded - nothing to do

![SteamTinkerLaunch_Nexus_download_failure](https://user-images.githubusercontent.com/1933471/210186776-6439b6c2-7cf1-4618-8590-ab6c9d77711a.png)

I'm launching MO2 in game mode.

**Comment by Lii-Kun** [comment]
I only skimmed the older posts so I'm sorry if this was said already. With version 12 of STL Vortex was handling the links fine(tho vortex has other issues). MO2 links wouldn't work if you opened it via the main menu of STL but would work if you hit play then opened MO2 via the MO2 popup.(I think a post earlier said the opposite worked).

**Comment by sonic2kk** [comment]
> MO2 links wouldn't work if you opened it via the main menu of STL but would work if you hit play then opened MO2 via the MO2 popup.(I think a post earlier said the opposite worked).

Historically MO2 links only reliably worked when opened in MO2 Game Mode i.e. the pop-up window you described (not Steam Deck Game Mode, that is separate). Since the MO2 Overhaul branch was merged though several users reported that the link handling worked regardless.

Thanks for confirming this though. The remaining issue seems to be on Steam Deck and affects ModOrganizer 2, Vortex and Hedge Mod Manager. The issue seems to be that it opens some kind of window asking for an application to handle the links, which I have not seen on desktop even when using Flatpaks. Instead of silently handling the links once the file association is set, or asking if you want the browser to open these applications, it asks which application to open to handle them. For some reason though when SteamTinkerLaunch uses `xdg-mime` to do this, it seems to confuse Firefox (and Chrome based on reports from users that unfortunately use Chrome on Steam Deck) and so nothing happens in some cases.

Though having said that, there hasn't been much activity on this issue in a while, so I guess the one-click links are basically unused (I personally dislike them anyway). I'm not sure what the path forward to fix the problem is, maybe there is no clean solution other than to just manually download your mods, which I would recommend all users to do anyway.

**Comment by har-nick** [comment]
Now I was going to test this with different browsers, but I'm sure this issue _knows_ that's happening lol

Installed Chrome, Edge, and Falkon (all Flatpak). Tested those, and they all worked fine. Tried Firefox Flatpak again, which didn't work 5 minutes ago, and it worked just fine.

I'm telling you, it _knows_.

**Comment by sonic2kk** [comment]
Boy, this issue really seems to be intermittent. I honestly don't know why it's so temperamental. It just wants to make a liar out of all of us :sweat_smile: 

I wonder if it's anything to do with the `NoDisplay` and/or `Hidden` sections of the `.desktop` file. I had a theory before that this made a difference on some systems but I don't remember the details of what i was thinking back then.

One day we'll get to the bottom of this! Unfortunately that day is not today...

**Comment by har-nick** [comment]
`Hidden` has an effect, as [the spec](https://specifications.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html) lists it as the equivalent of "deleting" the desktop file. I tested it set to `True` and nothing happened when starting a download. Setting it to `False` made it work as expected.

If the created desktop file is identical to the one in the MO2 wiki, which doesn't have a `Hidden` value set, I could see a distro assuming a value for it, but I doubt that's the issue, since it's affected myself regardless of having that value set.

**Comment by sonic2kk** [comment]
It would be interesting to see some tests between having various different combinations of `true` and `false for the `Hidden` and `NoDisplay` keys. Iwonder if some combination of these, or if simply specifying `Hidden=false` can fix the issue. I would be especially curious about this on Steam Deck, however I generally don't mod on my Steam Deck (I've already filled about 80% of my 512GB drive and my two 512gb microSD cards :sweat_smile:) so I can't test this personally.

You may be onto something for sure with the `Hidden` field in the `.desktop` file. I will test on my desktop when I have a chance with `Hidden=false` in the `.desktop` file as well. If testing goes well, we could experimentally write out `Hidden=false` to the `.desktop` file on its creation as my file doesn't have it at least and I don't think it's specified in the code. Updating this could be an issue though, perhaps there is a command to re-create the `.desktop` files.

Following this, the MO2 wiki could have its `.desktop` file examples updated as well :-) It should probably also be updated for local installs / Steam Deck installs, since the `/usr/bin/steamtinkerlaunch` path is not valid for every install.

**Comment by har-nick** [comment]
If I remember rightly, `NoDisplay` simply stops it appearing in the "Choose a desktop file to handle this link" menu that's given to you. You'd use it in conjunction with Firefox's Application menu to set a default desktop file handler and never see it again.

**Comment by BrandonKingM** [comment]
Personally I've been using the built in mod browser extension for the past year on Steam Deck. You can't download mods outside Nexus Mods, and you'll need to download mods you have to install manually outside this browser because it automatically downloads all regardless of what it is through Vortex but it's worked for me. I hope this issue does get solved though because I'd love to once again be able to open tabs when redownloading my mod history instead of the page refreshing to page 1 after each and every mod since this extension doesn't have tabs.

https://www.nexusmods.com/site/mods/88

**Comment by BrandonKingM** [comment]
>``
> NoDisplay=false
> Hidden=true
> ```

I just had it work for the first time I've ever tried it, having Chrome start a mod download to Vortex on my Steam Deck. I set these to 

>NoDisplay=false
> Hidden=false

The only issue I'm having is more just a small annoyance rather than a real issue, it claims that Vortex is the default program to open them, but every so often I click on open in Vortex i have to reselect it in the menu

**Comment by sonic2kk** [comment]
**Wow**! Okay, that is excellent news! In the near-ish future I will experimentally implement writing this out to the `.desktop` file on a branch and specifically request testing. I would like to see testing from other users across browsers, distros and on Steam Deck. I will also test myself with these fields before opening a branch to make sure it still works for me, which I would expect it to :-)

Great progress here @har-nick and @BrandonKingM, thank you very much!!

> but every so often I click on open in Vortex i have to reselect it in the menu

Does your mod download work after this though? Having to re-select is a slight annoyance but it could just be a browser quirk. I think we could live with this, probably :-)

**Comment by sonic2kk** [comment]
I created a new branch, `a-link-to-the-past`, in which I updated the `.desktop` file generation for MO2 and Vortex (and HMM) to have the following lines:

```
NoDisplay=false
Hidden=false
```

On master we have `NoDisplay=true`, and `Hidden` is entirely unspecified. If you install this branch of STL (ProtonUp-Qt with Advanced Mode enabled allows you to do this by installing STL-git and changing the branch) and then rename your existing MO2/Vortex `.desktop` file, it should re-create the next time you open the program.

Please feel free to test, anyone reading this issue, and report the behaviour. I have not tested yet but if this works as expected for a good portion of users who test, and if it does not break existing behaviour, I will be very happy and will look into merging.

One concern that I have is that this will cause Vortex/MO2 `.desktop` files (**which are only for link handling**) to show up in the Application Menu, which it **absolutely** not correct.

**Comment by har-nick** [comment]
Now I can't test this branch since I have Flatpak, but I've replicated the `.desktop` file to the same spec.

Fingers crossed this is the fix!

> One concern that I have is that this will cause Vortex/MO2 .desktop files (which are only for link handling) to show up in the Application Menu, which it absolutely not correct.

With `NoDisplay=false`, that's unfortunately what happens. As mentioned, I doubt that `NoDisplay` is the cause of this bug, but if it is, having an extra desktop entry in your XDG prompt is a tiny price to pay.

**Comment by cplr** [comment]
I tried the `a-link-to-the-past` branch, and can confirm I was able to open NXM links using Firefox (flatpak) on SteamOS 3.4 👍 

I had some weird hiccups where it would start the download, it would finish but then it would be "cancelled", and I would have to attempt a download again. And I also had to manually select the icon-less SteamTinkerLaunch from the list of "apps" to open the link, but after a few times that stopped popping up and it "just worked". Thanks!

_edit_: on Steam Deck is it recommended that you install MO2 as a Portable Instance or a Global Instance? Does it matter?

**Comment by sonic2kk** [comment]
Thank you for testing @cplr !! That is great news!

Is there an entry for MO2 in the application list? By default this is the Deck icon in the bottom left that lets you browse installed applications.

> I had some weird hiccups where it would start the download, it would finish but then it would be "cancelled", and I would have to attempt a download again. And I also had to manually select the icon-less SteamTinkerLaunch from the list of "apps" to open the link, but after a few times that stopped popping up and it "just worked". Thanks!

This is very strange, especially selecting the icon. This should ***not*** happen but it seems to happen on some perhaps weirdly configured distributions. I have never once been asked which application to use to handle links by my browser, but even in general this happens on SteamOS and I've heard reports that it happens on Pop!_OS. So it could be a configuration thing :woman_shrugging: I know the dialog you're referring to, but I have never seen it on any of my systems in the last ~5 years of using Linux for gaming.

The downloading issue is probably some network/wine weirdness, if it gets to downloading in MO2 then it isn't a SteamTinkerLaunch issue as far as I'm concerned and is likely related to Wine. Annoying, for sure, but not something that can be resolved from my side :-)

> on Steam Deck is it recommended that you install MO2 as a Portable Instance or a Global Instance? Does it matter?

There was some discussion previously that portable instances require more work. I don't think I've ever (at least intentionally) used a portable instance, I just use whatever the default is on PC (I don't mod games on Steam Deck where I can help it). For discussion on portable instances see #751 and this comment (https://github.com/sonic2kk/steamtinkerlaunch/issues/692#issuecomment-1356509825) discussing portable instances, though both of those may only apply to using the microSD card.

To me, and based on my understanding, global instances sound like they make the most sense to use, so I would stick with those no matter if you're using MO2 on Windows, Linux Desktop or Steam Deck. That's just my take though, I'm not a MO2 power user and input (and if necessary, feature implementations) will need to come from more experienced MO2 users to give a better answer I'm afraid.

<hr>

I will do some testing on my own Linux Desktop, and hopefully this issue can get some feedback from other Linux Desktop users, basically to give some confidence that this will be a go-forward solution. If updating these `.desktop` files works this should fix link handling for MO2, Vortex and HMM, which would be great. Given that you also had to select MO2 to handle links from your browser, I think that this strange dialog is what's causing the issues for users here: The browser, for some reason, expects to prompt users for which application to use, when really this shouldn't be the case and isn't the case outside of a handful of distros that unfortunately happens to include SteamOS.

As for the initial opening issue, where links did work but then stopped, I have no ideas on why that would've happened. But if the solution is as straightforward as what we've come up with here, hopefully this can be merged in the near-ish future. I'm hoping to roll a 12.12 release sometime in the near future and I'd like this to be one of the fixes in place.

**Comment by sonic2kk** [comment]
I put the branch up in a draft PR. I'll do some quick testing on my machines, and wait for further user testing. Hopefully it won't take too long to get more feedback. Once I'm confident it works for me and the users that tested I'll merge.

This issue has been open for a long time so I don't expect OP to respond, so I'll close this issue once merged. Usually I'd wait for OP's response if an isn't overly stale but this turned into more of a discussion around a solution than anything else so I think it's ok to close.

@SentaiBrad you're still welcome to test and give feedback of course but no worries either way, this issue is over 4 months old at this point 😅

**Comment by har-nick** [comment]
@cplr 

> I had some weird hiccups where it would start the download, it would finish but then it would be "cancelled"

Thankfully (or not), that's an MO2 issue. Have had it for years on Windows and Linux. Just delete your downloaded mod and redownload. Sometimes it'll also get stuck in the downloading phase, requiring you to cancel, delete, and redownload.

> And I also had to manually select the icon-less SteamTinkerLaunch from the list of "apps" to open the link, but after a few times that stopped popping up and it "just worked".

Sounds like your browser didn't have a default handler set. Can't say why it didn't set it to the default or why it's only done it now.

@sonic2kk While I'm not on MO2 very often, I can also confirm I've had zero issues since creating a new Desktop file.

**Comment by cplr** [comment]
Yeah, I mentioned that downloading issue more out of curiosity if others were seeing it. Definitely felt like a MO2 bug, and sounds like it is :).

Looking forward to this being on a main release! Is there a good way to upgrade to the regular train in Proton-Up-Qt or do you just remove the old and install the new?

**Comment by BrandonKingM** [comment]
Sent from my T-Mobile 5G Device
Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by sonic2kk** [comment]
Thanks for all the feedback everyone!

> Is there a good way to upgrade to the regular train in Proton-Up-Qt or do you just remove the old and install the new?

Could you explain what you mean @cplr ? If you mean how to update when this is merged into master, you can just install SteamTinkerLaunch-git. If you mean once it's in a release (e.g. 12.12 or something when it eventually comes out) just updating will suffice there as well, you can install from ProtonUp-Qt using the regular package.

@BrandonKingM I think something happened and the message didn't come through

**Comment by sonic2kk** [comment]
I should also point out that the branch has had some updates, if anyone is free to re-test that downloading links works I'd appreciate it! You can just rename your existing `.desktop` files likely located under `~/.local/share/applications`, and then run MO2/Vortex using the same branch `a-link-to-the-past` (branch naming is fun sometimes 😉). 

I made the following major changes yesterday:
- There is now an icon for the .desktop files, just the STL icon. This should make it identifiable if prompted for which application to use
- A warning dialog appears if you try to launch the .desktop file from the application menu, but this should not happen if you try to download a mod normally (it should only appear when the argument to the URL downloading function is blank, i.e. when `-z "$1"` is true). The dialog is to tell people that this is not the intended way to launch MO2, etc 
- The category for the entry should have changed to hopefully keep it out of sight a little bit

If all still works as expected then I'll probably merge this in when I have a chance in the next few days. It works on my machine but of course I didn't have this issue to begin with 😅 So testing from the community is always appreciated.

Thanks to everyone who helped Investigate and test, this issue would not have made any progress had it not been for your assistance here and I'll be sure to note this in the changelog.

**Comment by cplr** [comment]
I tried a fresh install of STL with the latest updates to the branch and everything is still working in top condition.

**Comment by sonic2kk** [comment]
Woo! Seems like my Sunday just got a lot better :sweat_smile: I will wait for more feedback but I am fairly confident in it now. I may make a couple of very minor adjustments over the next day or so but it should be merged in soon.

Once this is merged to master I'll add a note here. Providing there are no bug reports after merging that require reverting the merge this will also go into the next release, no ETA on when that is, I really just release when I feel like there are enough good fixes to go. Probably early next month I would say but no guarantees :wink: 

Thank you again for testing, it seems the fix was fairly straightforward all along!

**Comment by sonic2kk** [comment]
Decided to live life on the edge and merge it anyway, and then get feedback and resolve it afterwards. The changes from the branch are now in master!

Based on the positive feedback here from the branch, things should still be working fine now. Please feel free to re-test if you'd like :smile: 

I will leave this issue open for a little bit, but if no other problems crop up I will close it sometime in the future.

Thanks everyone!

**Comment by BrandonKingM** [comment]
> Decided to live life on the edge and merge it anyway, and then get feedback and resolve it afterwards. The changes from the branch are now in master!

This might be a stupid question but is the one on ProtonUp-Qt updated with the changes for Vortex? I can manually install it on my Deck if not. I love ProtonUp-Qt but i been using STL since long before it was available on that.

**Comment by BrandonKingM** [comment]
> @BrandonKingM I think something happened and the message didn't come through

I deleted it, must have been a misclick. I never use my email browser for things except to open it up in Chrome

**Comment by sonic2kk** [comment]
Seems resolved, closing :-)
